### PR TITLE
Add EventSource logging to CryptographyClient

### LINKED ISSUE #7648: Add EventSource logging to CryptographyClient
There are a few places places in code we should hook up trace logging to describe why we didn't use a local key (i.e. primarily because we have no corresponding crypto class from .NET).

### PR DESCRIPTION
**Pull request by heaths**
Fixes #7648

### CODE REVIEW

File: sdk/keyvault/Azure.Security.KeyVault.Keys/src/KeysEventSource.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        internal const int CryptographicExceptionEvent = 4;
+
+        private const string EventSourceName = "Azure-Security-KeyVault-Keys";
+
+        private KeysEventSource() : base(EventSourceName, EventSourceSettings.Default, AzureEventSourceListener.TraitName, AzureEventSourceListener.TraitValue) { }
+
+        public static KeysEventSource Singleton { get; } = new KeysEventSource();
+
+        [NonEvent]
+        public void AlgorithmNotSupported<T>(string operation, ref T algorithm) where T : struct
```
Comment (pakrym) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/src/KeysEventSource.cs]: is `ref T` overload required?

File: sdk/keyvault/Azure.Security.KeyVault.Shared/tests/Azure.Security.KeyVault.Shared.Tests/TestExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using NUnit.Framework;
+
+namespace Azure.Security.KeyVault.Tests
+{
+    internal static class TestExtensions
+    {
+        public static TestCaseData ConditionalIgnore(this TestCaseData source, bool condition, string reason)
```
Comment (pakrym) [inline sdk/keyvault/Azure.Security.KeyVault.Shared/tests/Azure.Security.KeyVault.Shared.Tests/TestExtensions.cs]: Why not just use an if in test data generation method?

**Review by pakrym — APPROVED** [review]
Nice, thank you for doing this and adding good tests!

File: sdk/keyvault/Azure.Security.KeyVault.Keys/src/KeysEventSource.cs (hunk 1 above)
Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/src/KeysEventSource.cs]: Any reason to pass it by value? Requires an extra allocation, albeit small for which structs are intended to be passed here.

Comment (pakrym) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/src/KeysEventSource.cs]: It's the same stack size and it's simpler.

File: sdk/keyvault/Azure.Security.KeyVault.Shared/tests/Azure.Security.KeyVault.Shared.Tests/TestExtensions.cs (hunk 2 above)
Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Shared/tests/Azure.Security.KeyVault.Shared.Tests/TestExtensions.cs]: The methods get ugly and harder to read/maintain. I already have more intended use for making a shared source test project anyway. Just figured this was enough of a proverbial straw to finally split the project like I should've in the first place when I added it.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/src/KeysEventSource.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            {
+                string message = FormatException(ex);
+                CryptographicException(operation, message);
+            }
+        }
+
+        [Event(CryptographicExceptionEvent, Level = EventLevel.Informational, Message = "A cryptographic exception occured: {1}.\r\nWill try the {0} operation in the service.")]
+        public void CryptographicException(string operation, string message) => WriteEvent(CryptographicExceptionEvent, operation, message);
+
+        private static string FormatException(Exception ex)
```
[the remark this answers is not in the document]
Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/src/KeysEventSource.cs]: Opened https://github.com/Azure/azure-sdk-for-net/issues/8297

**Pull request merged**